kura/org.eclipse.kura.linux.watchdog/src/main/java/org/eclipse/kura/linux/watchdog/WatchdogServiceImpl.java

```diff
@@ -206,118 +144,149 @@ public void unregisterCriticalService(CriticalComponent criticalComponent) {
+                if (csi.isTimedOut()) {
+                    failure = true;
+                    logger.warn("Critical service {} failed -> SYSTEM REBOOT", csi.getName());
+                    break;
                 }
             }
-        } else { // ! m_enabled
-            if (this.m_configEnabled) {
-                File f = null;
-                FileWriter bw = null;
+
+            // refresh watchdog if there aren't failures and the watchdog must not be stopped.
+            if (!failure && !this.watchdogToStop) {
+                refreshWatchdog();
+            } else {
                 try {
-                    f = new File(this.watchdogDevice);
-                    bw = new FileWriter(f);
-                    bw.write('w');
-                    bw.flush();
-                    bw.close();
-                    // m_watchdogToStart=false;
-                    this.m_enabled = true;
-                    s_logger.info("watchdog started");
-                } catch (IOException e) {
-                    s_logger.error("IOException on start watchdog", e);
-                } finally {
-                    try {
-                        if (bw != null) {
-                            bw.close();
-                        }
-                    } catch (IOException e) {
-                        s_logger.error("IOException on closing watchdog file", e);
-                    }
+                    performWatchdogActions();
+                } catch (KuraException e) {
+                    logger.error("Failed to perform watchdog actions.", e);
                 }
             }
+        } else {
+            if (this.configEnabled) {
+                refreshWatchdog();
+                this.isEnabled = true;
+            }
         }
     }
 
+    private synchronized void performWatchdogActions() throws KuraException {
+        // watchdogToStop is used to avoid multiple action executions and kill Kura after a while
+        if (!this.watchdogToStop) {
+            this.watchdogToStop = true;
+            runCommand("reboot");
+        } else {
+            refreshWatchdog();
+        }
+    }
+
+    private void runCommand(String command) throws KuraException {
+        try {
+            Process proc = Runtime.getRuntime().exec(command);
+            proc.waitFor();
+        } catch (IOException e) {
+            throw new KuraException(KuraErrorCode.OS_COMMAND_ERROR, e, "'" + command + "' failed");
+        } catch (InterruptedException e) {
+            throw new KuraException(KuraErrorCode.OS_COMMAND_ERROR, e, "'" + command + "' failed");
+        }
+    }
+
+    private void refreshWatchdog() {
+        writeWatchdogDevice("w");
+    }
+
     private void disableWatchdog() {
-        if (this.m_enabled) {
-            File f = null;
-            FileWriter bw = null;
+        if (this.isEnabled) {
+            writeWatchdogDevice("V");
+            this.isEnabled = false;
```

Comment:
It's used by the installer/upgrader to try to understand if the watchdog was enabled and refresh it if needed.